Add NS1 as a provider

mburtless opened the pull request:
As requested in #957 this PR adds a provider for NS1's DNS platform.  Please let me know what you think and if any modifications are required here.

mburtless commented:
Amended authorship on initial commit and linked my Linux Foundation ID user with GitHub

Raffo commented:
/check-cla

Raffo commented:
@mburtless Can you please sign the CLA?

mburtless commented:
@Raffo - Sorry, didn't realize I had do anything after the CCLA manager for my org whitelisted my email and I accepted their invite.  I think I've signed now.

Raffo commented on provider/ns1.go:
```diff
@@ -0,0 +1,295 @@
+
+	for _, zone := range zones {
+
+		// TODO handle Header Codes
```
What is missing here? Can we resolve this TODO instead of leaving it here?

mburtless replied:
This TODO is to handle the edge case where NS1 paginates the response.  In this case this would only occur when a zone contains 10,000+ records.  Unfortunately, this is not yet handled by the NS1 GO SDK, so to resolve this now would require manually inspecting the headers and making additional requests for the next page of records.

If this project is OK with it, I'd prefer to open a PR against ns1-go to add pagination support and then come back to add support for this in the provider.

Raffo commented on provider/ns1.go:
```diff
@@ -0,0 +1,295 @@
+
+// Zones returns the list of hosted zones.
+func (p *NS1Provider) zonesFiltered() ([]*dns.Zone, error) {
+	// TODO handle Header Codes
```
Same comment as before: can we resolve this TODO instead of leaving it here?

mburtless replied:
As before, this TODO is to handle the edge case where NS1 paginates the response.  In this case, this will only occur with 2500+ zones in a single account and can be mitigated by using an API key that only has access to a subset of zones.

Again, this would be cleaner to resolve after we get pagination supported via the ns1-go sdk

njuettner commented:
@mburtless we dropped `dep` and using `go mod` now. Can you solve the conflicts please?
Also please add tests for **types.go** and a tutorial how to use it, see [tutorials](https://github.com/kubernetes-incubator/external-dns/tree/master/docs/tutorials). You might also want to add it in the **README** file:
* https://github.com/kubernetes-incubator/external-dns#the-latest-release-v05
* https://github.com/kubernetes-incubator/external-dns#status-of-providers
* https://github.com/kubernetes-incubator/external-dns#deploying-to-a-cluster
* https://github.com/kubernetes-incubator/external-dns#v05---current-version

Than I'm more than happy to merge your PR 🙂.

mburtless commented:
@njuettner - Thanks for the feedback.  I've rebased the fork off of master and added our dependency to go.mod and go.sum.  I've also updated the README.  We're just finalizing the tutorial now, will add that in when complete.

In the meantime, can you let me know what additions you're looking for in `types_test.go`?  This provider doesn't actually add any new flags other than the new `ns1` value for the existing `provider` flag, so it's not obvious to me what I should be updating there.

njuettner commented:
/lgtm
/approve

njuettner commented:
@mburtless I didn't saw it correctly that you don't use any other types, all good. 
Thanks again for the PR 👍.
